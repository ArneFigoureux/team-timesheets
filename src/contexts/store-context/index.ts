export * from "./store-context";
export * from "./store-provider";
export * from "./use-store";
