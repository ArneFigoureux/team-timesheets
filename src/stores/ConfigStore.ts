import { observable, computed } from 'mobx';
import { Collection, ICollection } from "firestorable";
import { IRootStore } from './RootStore';
import { IProject, ITask, IClient, ITeam } from '../../common/dist';
import { firestore } from '../firebase/myFirebase';

export interface IConfigStore {
    projects: ICollection<IProject>;
    tasks: ICollection<ITask>;
    clientsCollection: ICollection<IClient>;
    teamsCollection: ICollection<IClient>;
    clients: (IClient & {id: string})[];
    teams: (ITeam & {id: string, isSelected: boolean})[];
    taskId?: string;
    projectId?: string;
    clientId?: string;
    teamId?: string;
}

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;

    readonly projects: ICollection<IProject>;
    readonly tasks: ICollection<IProject>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam>;

    @observable.ref taskId?: string;
    @observable.ref projectId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore, getCollection: (name: string) => firebase.firestore.CollectionReference) {
        // this._rootStore = rootStore;
        this.projects = observable(new Collection<IProject>(firestore, getCollection.bind(this, "projects"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.tasks = observable(new Collection<ITask>(firestore, getCollection.bind(this, "tasks"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.clientsCollection = observable(new Collection<ITask>(firestore, getCollection.bind(this, "clients"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.teamsCollection = observable(new Collection<ITeam>(firestore, getCollection.bind(this, "teams"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }
}