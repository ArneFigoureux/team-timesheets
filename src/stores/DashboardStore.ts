import { observable, reaction, computed, IObservableValue, action } from "mobx";
import { ICollection, Collection } from "../Firestorable/Collection";

import { IRootStore } from "./RootStore";
import { IRegistration, IRegistrationData, IGroupedRegistrations } from "./TimesheetsStore";
import * as deserializer from '../serialization/deserializer';
import * as serializer from '../serialization/serializer';

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration>;

    @observable private startDate: IObservableValue<Date | undefined> = observable.box();
    @observable private endDate: IObservableValue<Date | undefined> = observable.box();
    @observable.ref private userFilterValueField: string | undefined = undefined;
    @observable.ref private taskFilterValueField: string | undefined = undefined;
    @observable.ref private projectFilterValueField: string | undefined = undefined;

    constructor(rootStore: IRootStore) {
        this.registrationsField = observable(new Collection<IRegistration, IRegistrationData>(() => rootStore.getCollection("registrations"),
            {
                realtime: true,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration
            }));

        const updateRegistrationQuery = () => {
            this.registrationsField.query = ref =>
                this.userIdFilter(
                    this.projectIdFilter(
                        this.endDateFilter(
                            this.startDateFilter(
                                ref.where("deleted", "==", false)
                            )
                        )
                    )
                );
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.startDate, updateRegistrationQuery);
        reaction(() => this.endDate, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);

        this.registrationsField.getDocs();
    }

    @computed
    private get startDateFilter() {
        const startDate = this.startDate.get()
        return (query: firebase.firestore.Query) => startDate ? query.where("date", ">=", startDate) : query;
    }

    @computed
    private get endDateFilter() {
        const endDate = this.endDate.get()
        return (query: firebase.firestore.Query) => endDate ? query.where("date", "<=", endDate) : query;
    }

    @computed
    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (query: firebase.firestore.Query) => userId ? query.where("userId", "==", userId) : query;
    }

    @computed
    private get projectIdFilter() {
        return (query: firebase.firestore.Query) => this.projectFilterValue ? query.where("project", "==", this.projectFilterValue) : query;
    }

    @computed
    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    @computed
    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    @computed
    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    @computed
    private get registrations() {
        return Array.from(this.registrationsField.docs.values())
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    @action
    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    @action
    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    @action
    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}