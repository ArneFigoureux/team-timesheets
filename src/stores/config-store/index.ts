export * from "./config-store";
export * from "./use-config-store";