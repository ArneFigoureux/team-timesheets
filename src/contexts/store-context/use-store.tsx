import React from "react";
import { StoreContext } from "./store-context";

export const useStore = () => React.useContext(StoreContext);