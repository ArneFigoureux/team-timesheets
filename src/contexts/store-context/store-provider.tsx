import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store } from "../../stores/root-store";

export const StoreProvider = ({ children }: React.PropsWithChildren<{}>) => {
    const firebase = useFirebase();

    const auth = firebase.auth();
    const firestore = firebase.firestore();
    const storage = firebase.storage();

    const store = (window as any)["store"] = new Store({
        auth,
        firestore,
        storage,
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
