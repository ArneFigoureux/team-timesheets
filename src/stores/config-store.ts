import { observable, computed } from 'mobx';
import { Collection, ICollection, RealtimeMode, FetchMode } from "firestorable";
import { IRootStore } from './root-store';
import { firestore } from '../firebase/my-firebase';
import { IProject, ITask, IClient, IClientData, ITeam, ITeamData, ITaskData, IConfig, ConfigValue } from '../../common/dist';

import * as serializer from '../../common/serialization/serializer';
import * as deserializer from '../../common/serialization/deserializer';
import { LoginProvider } from '../firebase/types';

export type Configs = {
    // authClientId: string;
    loginProviders: LoginProvider[];
}

export interface IConfigStore extends ConfigStore {};

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;
    readonly tasks: ICollection<IProject, ITaskData>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam, ITeamData>;
    readonly configsCollection: Collection<IConfig>;

    @observable.ref taskId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore) {
        // this._rootStore = rootStore;
        const deps = { logger: console.log };

        this.tasks = new Collection<ITask, ITaskData>(
            firestore,
            "tasks",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.clientsCollection =
            new Collection<IClient, IClientData>(
                firestore,
                "clients",
                {
                    realtimeMode: RealtimeMode.on,
                    fetchMode: FetchMode.once,
                    query: ref => ref.orderBy("name_insensitive"),
                    serialize: serializer.convertNameWithIcon,
                    deserialize: deserializer.convertNameWithIcon,
                },
                deps,
            );

        this.teamsCollection = new Collection<ITeam, ITeamData>(
            firestore,
            "teams",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.configsCollection = new Collection<IConfig>(
            firestore,
            "configs",
            {
                fetchMode: FetchMode.once,
            }
        )
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }

    // To investigate: does getConfigValue needs mobx @computed attribute?
    public getConfigValue<T>(key: string): T;
    public getConfigValue<T>(key: string, isRequired: false): T | undefined;
    public getConfigValue<T extends ConfigValue>(key: string, isRequired = true): T | undefined {
        const doc = this.configsCollection.docs.find(c => c.data!.key === key);

        if (!doc) {
            if (isRequired) throw new Error(`Required config '${key}' is missing.`);
            else return undefined;
        }

        return doc.data!.value as unknown as T;
    }
}
