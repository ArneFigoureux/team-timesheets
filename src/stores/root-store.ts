import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { firestore } from "../firebase/my-firebase";
import { IProjectStore, ProjectStore } from "./project-store";

export interface IRootStore {

    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly getCollection: (name: string) => firebase.firestore.CollectionReference;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;

    public readonly getCollection: (name: string) => firebase.firestore.CollectionReference;

    constructor(getCollection: (name: string) => firebase.firestore.CollectionReference) {
        this.getCollection = getCollection;

        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this, getCollection);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);

    }
};

const store = (window as any)["store"] = new Store(name => firestore.collection(name));

export default store;

