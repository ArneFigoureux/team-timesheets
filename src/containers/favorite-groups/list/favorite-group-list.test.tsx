import React from "react";
import fs from "fs";
import path from "path";

import { clearFirestoreData, loadFirestoreRules, initializeTestApp } from "@firebase/rules-unit-testing";

import type firebase from "firebase";
import { render, waitFor } from "@testing-library/react";
import { Store } from "../../../stores/root-store";
import { useStore } from "../../../contexts/store-context";
import FavoriteGroupList from ".";

jest.mock("../../../contexts/store-context");

const projectId = "favorite-group-detail-page";

const app = initializeTestApp({
    projectId,
});

let store: Store;
const userId = "user-1";
let favoriteGroupIds: string[] = ["fav-1", "fav-2", "fav-3"];

const setupAsync = async () => {
    store = new Store({
        firestore: app.firestore(),
    });

    await store.auth.addDocument({
        uid: userId,
        divisionId: "",
        name: "User 1",
        recentProjects: [],
        roles: { user: true },
        tasks: new Map(),
        email: "user@timesheets.com",
    }, userId);

    store.auth.setUser({
        uid: userId,
    } as firebase.User);
};

beforeAll(async () => {
    await loadFirestoreRules({
        projectId,
        rules: fs.readFileSync(path.resolve(__dirname, "../../../../firestore.rules.test"), "utf8"),
    });
});

beforeEach(async () => {
    await setupAsync();
    (useStore as jest.Mock<ReturnType<typeof useStore>>).mockReturnValue(store);
});

afterEach(async () => {
    store.dispose();
    await clearFirestoreData({
        projectId,
    });
});

afterAll(() => app.delete());
describe("FavoriteGroupList", () => {
    describe("when there are no favorite groups", () => {
        it("should render an empty list", async () => {
            const {
                asFragment,
                unmount,
            } = render(
                <FavoriteGroupList />
            );

            await waitFor(() => expect(store.favorites.collection.isFetched).toBeTruthy());

            expect(asFragment()).toMatchSnapshot();

            unmount();
        });

        it("should rerender the list when favorite groups are added in the database in the background", async () => {
            const {
                asFragment,
                getByText,
                unmount,
            } = render(
                <FavoriteGroupList />
            );

            await waitFor(() => expect(store.favorites.collection.isFetched).toBeTruthy());

            expect(asFragment()).toMatchSnapshot();

            await store.favorites.addDocuments([
                {
                    name: "new fav group 1",
                    userId,
                },
            ]);

            await waitFor(() => getByText("new fav group 1"));

            unmount();
        });
    });

    describe("when there are favorite groups", () => {
        beforeEach(async () => {
            await Promise.all(
                [
                    store.favorites.addDocument(
                        {
                            name: "fav 1",
                            userId: userId,
                        },
                        "fav-1",
                    ),
                    store.favorites.addDocument(
                        {
                            name: "fav 2",
                            userId: userId,
                        },
                        "fav-2",
                    ),
                    store.favorites.addDocument(
                        {
                            name: "fav 3",
                            userId: userId,
                        },
                        "fav-3",
                    ),
                ]);

            await store.favorites.favoriteCollection.addAsync([
                {
                    groupId: favoriteGroupIds[1],
                    userId,
                    description: "fav reg 1",
                    time: 3,
                },
                {
                    groupId: favoriteGroupIds[1],
                    userId,
                    description: "fav reg 2",
                    time: 1,
                },
            ]);
        });

        it(
            "should render list of favorite groups",
            async () => {

                const {
                    getByText,
                    unmount,
                } = render(
                    <FavoriteGroupList />
                );

                await waitFor(() => {
                    getByText("fav 1");
                    getByText("fav 2");
                    getByText("fav 3");
                });

                unmount();
            },
        );

        it(
            "should rerender list when one of the groups is updated in the database in the background",
            async () => {

                const {
                    getByText,
                    unmount,
                } = render(
                    <FavoriteGroupList />
                );

                await waitFor(() => {
                    getByText("fav 1");
                    getByText("fav 2");
                    getByText("fav 3");
                });

                await store.favorites.collection.updateAsync(
                    {
                        name: "fav 2 bis",
                    },
                    "fav-2",
                );

                await waitFor(() => {
                    getByText("fav 1");
                    getByText("fav 2 bis");
                    getByText("fav 3");
                });

                unmount();
            },
        );

    });
});
