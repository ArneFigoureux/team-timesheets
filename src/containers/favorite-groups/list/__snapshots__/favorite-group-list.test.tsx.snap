// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteGroupList when there are no favorite groups should render an empty list 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <p>
      You don't have any favorites yet. Start adding registrations as favorite from your timesheet.
    </p>
    <p>
      Select registrations with the checkbox and click the favorite icon in the top action bar.
    </p>
  </div>
</DocumentFragment>
`;

exports[`FavoriteGroupList when there are no favorite groups should rerender the list when favorite groups are added in the database in the background 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <p>
      You don't have any favorites yet. Start adding registrations as favorite from your timesheet.
    </p>
    <p>
      Select registrations with the checkbox and click the favorite icon in the top action bar.
    </p>
  </div>
</DocumentFragment>
`;
